### ISSUE TITLE
May be a BUG in 'Conv2d' with "Padding_mode='circular', Padding=1"

### ISSUE THREAD

**Issue by Dezreal** [pytorch/pytorch#20981]
## 🐛 Bug

When 'Padding_mode' is 'circular', ‘Padding=1’ and 'Padding=0' are no different.
(In fact, I am not sure whether this is a bug or a special setting.)

## To Reproduce

```python
import torch
from torch import nn

m = nn.Conv2d(1, 1, 1, 1, padding=1, padding_mode='circular')
x = torch.rand(1, 1, 5, 5)

print(m(x).shape)
```

## Expected behavior

As code said:
```python
            expanded_padding = ((self.padding[1] + 1) // 2, self.padding[1] // 2,
                                (self.padding[0] + 1) // 2, self.padding[0] // 2)
```
the size of output is expected to equals to [1, 1, 6, 6], but the result is [1, 1, 5, 5]

## Environment

 - PyTorch Version: 1.1.0
 - OS: Win 10
 - pip
 - Python version: 3.6.7
 - CUDA/cuDNN version: None

## Additional context
If this is just a bug, then I think the problem is here: 
line 3066 in [functional.py](https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py)
Because of: `-(padding[-1] + padding[-2]):0` === 0

**Comment by qizhuli** [comment]
I also believe there is a bug with the circular padding; currently, directly replacing zero-padded convs with circular-padded convs leads to output of different spatial dimensions (if `kernel_size`>1). I don't think this should be the intended effect. See demo below:
```python
from torch import nn

x = torch.rand(1, 16, 32, 32)

filter_pzeros = nn.Conv2d(16, 16, 3, padding=1, padding_mode='zeros')
filter_pcircular = nn.Conv2d(16, 16, 3, padding=1, padding_mode='circular')

print(filter_pzeros(x).shape)
print(filter_pcircular(x).shape)
```
outputs:
```python
torch.Size([1, 16, 32, 32])
torch.Size([1, 16, 30, 30])
```
With `padding`=1 and `kernel_size`=3, the conv should preserve the input spatial dimension. The output from circular-padded conv is not right.

### Observations
1. The `expanded_padding` is unsymmetrical:
```python
expanded_padding = ((self.padding[1] + 1) // 2, self.padding[1] // 2,
                                (self.padding[0] + 1) // 2, self.padding[0] // 2)

```
With `padding`=1, we get `expanded_padding=(1, 0, 1, 0)`. This pads `x` to a size of `(1, 16, 33, 33)`. After conv with `kernel_size`=3, this would result in an output tensor of size `(1, 16, 31, 31)`. Shouldn't we pad (1, 1, 1, 1) evenly on all four directions (left, right, top, bottom), so that the spatial size would be preserved?

2. Even after taking into account the unsymmetrical `expanded_padding` implementation, the expected output size (`(1, 16, 31, 31)`) still does not match the actual output size of circular-padded conv (`(1, 16, 30, 30)`). After some further investigation, I realised that `circular` padding itself seems bugged which is illustrated by the following demo:
```python
from torch.nn import functional as F

x = torch.rand(1, 16, 32, 32)

y_pconstant = F.pad(x, (1, 0, 1, 0), mode='constant')
y_preplicate = F.pad(x, (1, 0, 1, 0), mode='replicate')
y_preflect = F.pad(x, (1, 0, 1, 0), mode='reflect')
y_pcircular = F.pad(x, (1, 0, 1, 0), mode='circular')

print(y_pconstant.shape)
print(y_preplicate.shape)
print(y_preflect.shape)
print(y_pcircular.shape)
```
 outputs:

```python
torch.Size([1, 16, 33, 33])
torch.Size([1, 16, 33, 33])
torch.Size([1, 16, 33, 33])
torch.Size([1, 16, 32, 32])
```
For some reason, `circular` padding ignored the padding instruction, and decided not to pad at all!

### TL;DR
We should take a look into `expanded_padding` in conv implementation, and `nn.functional.pad` using `mode`='circular'.

### Environment
- PyTorch version: 1.2.0

**Comment by Germanunkol** [comment]
I get similar behavior with nn.Conv3d. As soon as padding_mode is set to "circular", the tensor size is reduced by the convolution, even if kernel_size is 3 and padding is 1.

**Comment by sunyuanxi** [comment]
According to my experiment:
Inputs:
```
x = torch.rand(1, 16, 32, 32)

y_pcircular = F.pad(x, (0, 1, 0, 1), mode='circular')
print(y_pcircular.shape)
```
Outputs:
```
torch.Size([1, 16, 33, 33])
```
Inputs:
```
y_pcircular = F.pad(x, (1, 0, 1, 0), mode='circular')
print(y_pcircular.shape)
```
Outputs:
```
torch.Size([1, 16, 32, 32])
```

Therefore, I think that there's a bug in the GitHub source code for circular padding (```nn/functional.py``` line 3104 to line 3124): 
```
    input = torch.cat([input, input[:, :, 0:padding[-1]]], dim=2)
    input = torch.cat([input[:, :, -(padding[-1] + padding[-2]):-padding[-1]], input], dim=2)
```
It doesn't take into account the situation when ```padding[-1] = 0```.
If ```padding[-1] = 0```, then:
```
input[:, :, -(padding[-1] + padding[-2]):-padding[-1]] = input[:, :, -padding[-2]:0]
```
will select nothing, and thus lead to the error in circular padding. And samething for ```padding[-3]``` and ```padding[-5]```.

So maybe we can do something like:
```
if padding[-1] > 0:
    input = torch.cat([input, input[:, :, 0:padding[-1]]], dim=2)
    input = torch.cat([input[:, :, -(padding[-1] + padding[-2]):-padding[-1]], input], dim=2)
else:
    if padding[-2] > 0:
        input = torch.cat([input[:, :, -padding[-2]:], input], dim=2)
```

**Comment by sAviOr287** [comment]
Hi,

This issue still hasnt been fixed in torch 1.4 it seems like.

Are there any updates on this?

**Comment by VRandme** [comment]
I get similar behavior with nn.Conv1d. I am seeing tensor size reduction when it shouldn't

**Comment by sAviOr287** [comment]
You can solve it somehow for the nn.conv2D at least by creating a padding class:

`class padding_cir(nn.Module):

	'''
	Creating circular padding
	'''
	def __init__(self):
		super(padding_cir, self).__init__()

	def forward(self, x):
		# creating the circular padding
		return F.pad(x, (1, 1, 1, 1), mode='circular')`

and then when you create you list of layers you just dont have padding for you conv2d
`conv2d = nn.Conv2d(in_channels, v, kernel_size=3, bias=False)`
`layers = [padding_cir(), conv2d, self.activation]`
`nn.Sequential(*layers)`

**Comment by macknever** [comment]
why not (1,1,7,7)

**Comment by sAviOr287** [comment]
It worked with (1,1,1,1) for me

**Comment by mruberry** [comment]
We recently fixed circular padding and this issues appears to have been fixed. Would you please verify, @dezreal?
